File: GurmukhiFlutter/lib/models/user_preferences.dart
import 'app_language.dart';
import 'source_type.dart';

class UserPreferences {
  UserPreferences({
    required this.hasCompletedOnboarding,
    required this.preferredLanguage,
    required this.selectedSources,
    required this.notificationHour,
    required this.notificationMinute,
    required this.notificationTimes,
    required this.notificationsEnabled,
    required this.notificationTimesPerDay,
    required this.notificationUseRandomTimes,
  });

  bool hasCompletedOnboarding;
  AppLanguage preferredLanguage;
  List<SourceType> selectedSources;
  int notificationHour;
  int notificationMinute;
  List<int> notificationTimes;
  bool notificationsEnabled;
  int notificationTimesPerDay;
  bool notificationUseRandomTimes;

  static UserPreferences defaults() {
    return UserPreferences(
      hasCompletedOnboarding: false,
      preferredLanguage: AppLanguage.english,
      selectedSources: SourceType.values,
      notificationHour: 7,
      notificationMinute: 0,
      notificationTimes: const [420],
      notificationsEnabled: false,
      notificationTimesPerDay: 1,
      notificationUseRandomTimes: false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'hasCompletedOnboarding': hasCompletedOnboarding,
      'preferredLanguage': preferredLanguage.value,
      'selectedSources': selectedSources.map((e) => e.value).toList(),
      'notificationHour': notificationHour,
      'notificationMinute': notificationMinute,
      'notificationTimes': notificationTimes,
      'notificationsEnabled': notificationsEnabled,
      'notificationTimesPerDay': notificationTimesPerDay,
      'notificationUseRandomTimes': notificationUseRandomTimes,
    };
  }

  factory UserPreferences.fromJson(Map<String, dynamic> json) {
    final selected = (json['selectedSources'] as List<dynamic>? ?? const <dynamic>[])
        .map((item) => SourceType.fromValue(item.toString()))
        .toList();

    return UserPreferences(
      hasCompletedOnboarding: json['hasCompletedOnboarding'] as bool? ?? false,
      preferredLanguage: AppLanguage.fromValue(json['preferredLanguage'] as String? ?? ''),
      selectedSources: selected.isEmpty ? SourceType.values : selected,
      notificationHour: (json['notificationHour'] as int?) ?? 7,
      notificationMinute: (json['notificationMinute'] as int?) ?? 0,
      notificationTimes: ((json['notificationTimes'] as List<dynamic>? ?? const [420])
          .map((item) => (item as num).toInt())
          .toList()),
      notificationsEnabled: json['notificationsEnabled'] as bool? ?? false,
      notificationTimesPerDay: ((json['notificationTimesPerDay'] as int?) ?? 1).clamp(1, 5),
      notificationUseRandomTimes: json['notificationUseRandomTimes'] as bool? ?? false,
    );
  }
}

File: Models/UserPreferences.swift
import Foundation

struct UserPreferences: Codable {
    var hasCompletedOnboarding: Bool
    var preferredLanguage: AppLanguage
    var selectedSources: [SourceType]
    var notificationHour: Int
    var notificationMinute: Int
    var notificationTimes: [Int]
    var notificationsEnabled: Bool
    var notificationTimesPerDay: Int
    var notificationUseRandomTimes: Bool

    static let `default` = UserPreferences(
        hasCompletedOnboarding: false,
        preferredLanguage: .english,
        selectedSources: SourceType.allCases,
        notificationHour: 7,
        notificationMinute: 0,
        notificationTimes: [7 * 60],
        notificationsEnabled: false,
        notificationTimesPerDay: 1,
        notificationUseRandomTimes: false
    )

    private enum CodingKeys: String, CodingKey {
        case hasCompletedOnboarding
        case preferredLanguage
        case selectedSources
        case notificationHour
        case notificationMinute
        case notificationTimes
        case notificationsEnabled
        case notificationTimesPerDay
        case notificationUseRandomTimes
    }

    init(
        hasCompletedOnboarding: Bool,
        preferredLanguage: AppLanguage,
        selectedSources: [SourceType],
        notificationHour: Int,
        notificationMinute: Int,
        notificationTimes: [Int],
        notificationsEnabled: Bool,
        notificationTimesPerDay: Int,
        notificationUseRandomTimes: Bool
    ) {
        self.hasCompletedOnboarding = hasCompletedOnboarding
        self.preferredLanguage = preferredLanguage
        self.selectedSources = selectedSources
        self.notificationHour = notificationHour
        self.notificationMinute = notificationMinute
        self.notificationTimes = notificationTimes
        self.notificationsEnabled = notificationsEnabled
        self.notificationTimesPerDay = notificationTimesPerDay
        self.notificationUseRandomTimes = notificationUseRandomTimes
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        hasCompletedOnboarding = try container.decode(Bool.self, forKey: .hasCompletedOnboarding)
        preferredLanguage = try container.decodeIfPresent(AppLanguage.self, forKey: .preferredLanguage) ?? .english
        selectedSources = try container.decode([SourceType].self, forKey: .selectedSources)
        notificationHour = try container.decode(Int.self, forKey: .notificationHour)
        notificationMinute = try container.decode(Int.self, forKey: .notificationMinute)
        let fallbackMinute = max(0, min(1439, notificationHour * 60 + notificationMinute))
        let decodedTimes = try container.decodeIfPresent([Int].self, forKey: .notificationTimes) ?? [fallbackMinute]
        notificationTimes = decodedTimes
            .map { max(0, min(1439, $0)) }
            .sorted()
        notificationsEnabled = try container.decode(Bool.self, forKey: .notificationsEnabled)
        notificationTimesPerDay = max(1, min(5, try container.decodeIfPresent(Int.self, forKey: .notificationTimesPerDay) ?? 1))
        notificationUseRandomTimes = try container.decodeIfPresent(Bool.self, forKey: .notificationUseRandomTimes) ?? false
    }
}
